File: leetcode/p0133/Solution.kt
/**
 * Definition for a Node.
 * class Node(var `val`: Int) {
 *     var neighbors: ArrayList<Node?> = ArrayList<Node?>()
 * }
 */

class Solution {
    fun cloneGraph(node: Node?): Node? {
        if (node == null) {
            return null
        }
        val map = mutableMapOf<Int, Node>()
        val queue = ArrayDeque<Node>()
        val visitedSet = mutableSetOf<Node>()
        
        visitedSet.add(node)
        queue.add(node)
        while (!queue.isEmpty()) {
            val cur = queue.poll()
            //println(cur.`val`)
            
            val copied = map[cur.`val`] ?: Node(cur.`val`)
            map[cur.`val`] = copied
            
            for (neighbor in cur.neighbors) {
                if (neighbor == null) {
                    continue
                }
                
                val tmp = map[neighbor.`val`] ?: Node(neighbor.`val`)
                map[neighbor.`val`] = tmp
                
                //if (!copied.neighbors.contains(tmp)) {
                    copied.neighbors.add(tmp)
                //}
                
                if (visitedSet.contains(neighbor)) {
                    continue
                }
                
                visitedSet.add(neighbor)
                queue.offer(neighbor)
            }
            
        }
        return map[1]
    }
    
}

File: leetcode/p0133/Solution.swift
/**
 * Definition for a Node.
 * public class Node {
 *     public var val: Int
 *     public var neighbors: [Node?]
 *     public init(_ val: Int) {
 *         self.val = val
 *         self.neighbors = []
 *     }
 * }
 */

class Solution {
    func cloneGraph(_ node: Node?) -> Node? {
        var map = [Int:Node]()
        return helper(node, &map)
    }
    private func helper(_ node: Node?, _ map: inout [Int:Node]) -> Node? {
        //print("helper \(node?.val ?? nil)")
        if let node = node {
            if let cached = map[node.val] {
                return cached
            }
            var copy = Node(node.val)
            map[node.val] = copy
            for neighbor in node.neighbors {
                copy.neighbors.append(helper(neighbor, &map))
            }
            return copy
        } else {
            return nil
        }
    }
}
